src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSignature.cs

```diff
@@ -0,0 +1,254 @@
+            _encodedSignProtectedHeaders = encodedSignProtectedHeaders;
+            _signature = signature;
+        }
+
+        public CoseHeaderMap ProtectedHeaders { get; }
+        public CoseHeaderMap UnprotectedHeaders { get; }
+
+        private CoseMultiSignMessage? _message;
+        internal CoseMultiSignMessage Message
+        {
+            get
+            {
+                Debug.Assert(_message != null);
+                return _message;
+            }
+            set
+            {
+                _message = value;
+            }
+        }
+
+        public bool VerifyEmbedded(AsymmetricAlgorithm key, ReadOnlySpan<byte> associatedData)
+        {
+            if (key is null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (Message.IsDetached)
+            {
+                throw new CryptographicException(SR.Sign1VerifyContentWasDetached);
+            }
+
+            return VerifyCore(key, Message.Content.Value.Span, null, associatedData, CoseHelpers.GetKeyType(key));
+        }
+
+        public bool VerifyEmbedded(AsymmetricAlgorithm key, byte[]? associatedData = null)
+        {
+            if (key is null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (Message.IsDetached)
+            {
+                throw new CryptographicException(SR.Sign1VerifyContentWasDetached);
+            }
+
+            return VerifyCore(key, Message.Content.Value.Span, null, associatedData, CoseHelpers.GetKeyType(key));
+        }
+
+        public bool VerifyDetached(AsymmetricAlgorithm key, byte[] detachedContent, byte[]? associatedData = null)
+        {
+            if (key is null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (detachedContent is null)
+            {
+                throw new ArgumentNullException(nameof(detachedContent));
+            }
+
+            if (!Message.IsDetached)
+            {
+                throw new CryptographicException(SR.Sign1VerifyContentWasEmbedded);
+            }
+
+            return VerifyCore(key, detachedContent, null, associatedData, CoseHelpers.GetKeyType(key));
+        }
+
+        public bool VerifyDetached(AsymmetricAlgorithm key, ReadOnlySpan<byte> detachedContent, ReadOnlySpan<byte> associatedData = default)
+        {
+            if (key is null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (!Message.IsDetached)
+            {
+                throw new CryptographicException(SR.Sign1VerifyContentWasEmbedded);
+            }
+
+            return VerifyCore(key, detachedContent, null, associatedData, CoseHelpers.GetKeyType(key));
+        }
+
+        public bool VerifyDetached(AsymmetricAlgorithm key, Stream detachedContent, ReadOnlySpan<byte> associatedData = default)
```

Comment:
In `VerifyDetachedAsync`, we throw if `detachedContent` cannot read. I don't see that same check in the synchronous version.